### PR TITLE
Add config to manage how long logs are stored

### LINKED ISSUE #2942: Can you please add logs lifetime config option?
Hello, can you please add logs_lifetime config option? That you can set time (in ms) that how long logs will be stored in database. 
After that time remove logs from database.

By default this option should be disabled (seted to 0 = dont delete logs)

### PR DESCRIPTION
**Pull request by Atharv-Agarwal**
Closes #2942 

This pr adds a config option called `log_expiration` which can be configured in the same way as `thread_cooldown` etc. 
According to the `closed_at` field this config option will delete logs which have passed the time specified in `log_expiration`. 

This will work retrospectively.

### CODE REVIEW

**Comment by Cordila** [comment]
I have tested this extensively but it would be better if someone could review this to make sure its efficient. Also since setting the config option will immediately start deleting old logs not sure if we should have some kind of safeguard?

**Comment by StephenDaDev** [comment]
I think a safeguard is probably a good measure just to make sure nobody does anything they'd rather not do

**Comment by Cordila** [comment]
> I think a safeguard is probably a good measure just to make sure nobody does anything they'd rather not do

Actually if you look at it, the config option is set after the bot is running, they will need to restart the bot for it to come into effect. In this time they can remove the config option to prevent logs from being deleted. However vise versa if the logs are being deleted and they remove the config option, the deletion will stop without needing a restart. 

Maybe this is enough?

**Comment by StephenDaDev** [comment]
@lorenzo132 What do you think?

**Comment by lorenzo132** [comment]
> @lorenzo132 What do you think?

I'm a fan of the logs expiration config, specialy on a company wise level this is a very nice feature to have.

**Comment by Cordila** [comment]
Used `delete_many` instead of looping `delete_one` based on recommendation by @khakers

File: bot.py [hunk 1]
```diff
@@ -1714,6 +1712,41 @@ async def before_autoupdate(self):
             self.autoupdate.cancel()
             return
 
+    @tasks.loop()
+    async def log_expiry(self):
```
Comment (khakers) [inline bot.py:1716]: Not a showstopper, but this method seems incredibly overcomplicated for what it *needs* to do. I would say it could be trimmed to down to deleting the logs at line 1723 (you get the amount deleted from this so you can log it) and then repeating again after a fixed time, which is how these tasks are typically done anyways. Doing this would make the code much more readable and efficient (though I don't think the extra database calls this currently has would ever actually matter at the scale it's likely to be used at).

Comment (Cordila) [reply bot.py:1716]: I don't think it would make it more efficient because instead of running every x minutes and not having any logs to delete, the task will only run when a log needs to be deleted.

Comment (khakers) [reply bot.py:1716]: It's probably going to be mostly irrelevant in efficiency either way (running one query on the database once an hour or day or something that only returns deletion data is basically meaningless, as are basically all the other ones here). I was mostly referring to code readability that you would get by removing all the exit points in this method (and basically removing most of the code) and not duplicating the same query multiple times.

For reference, this is what I'm talking about (untested code):

```python
async def expire_logs(self, log_expire_after):
        now = datetime.utcnow()
        expiration_datetime = now - log_expire_after
        expired_logs = await self.db.logs.delete_many({"closed_at": {"$lte": str(expiration_datetime)}})

        logger.info(f"Deleted {len(expired_logs.deleted_count)} expired logs.")
 ```

imo this is much easier to read and the only functionality you lose is precise scheduling (and you gain being able to easily run it with a command if you wanted)

Comment (Taaku18) [reply bot.py:1716]: I agree that the majority of the function looks overengineered. I think you can use a constant period loop over the function, and then directly delete from the database based on a query would work better and be more efficient.

We can have the task auto-repeat every hour:
`@tasks.loop(hours=1, reconnect=False)`

File: bot.py [hunk 2]
```diff
@@ -1714,6 +1712,41 @@ async def before_autoupdate(self):
...
+        log_expire_after = self.config.get("log_expiration")
+        if log_expire_after == isodate.Duration():
+            return self.log_expiry.stop()
+
+        now = datetime.utcnow()
```
Comment (Taaku18) [inline bot.py:1721]:
```suggestion
        now = discord.utils.utcnow()
```

For consistency with the rest of the codebase, we use discord.py's `utcnow()`, which includes the timezone info.

File: bot.py [hunk 3]
```diff
@@ -1714,6 +1712,41 @@ async def before_autoupdate(self):
...
+        if not expired_logs:
+            fetch = await self.db.logs.find().sort("closed_at", 1).limit(1).to_list(None)
+            if fetch:
+                for x in fetch:
+                    if x["closed_at"] > str(expiration_datetime):
+                        closed_at = datetime.strptime(x["closed_at"], "%Y-%m-%d %H:%M:%S.%f%z")
+                        next_expiry = closed_at + log_expire_after
+                        return await discord.utils.sleep_until(next_expiry)
+
+        await self.db.logs.delete_many({"closed_at": {"$lte": str(expiration_datetime)}})
```
Comment (Taaku18) [inline bot.py:1739]: `closed_at` is stored as a string, see #3115. You cannot directly compare `closed_at` using `lte` string comparison. You'll need to use an MongoDB aggregation pipeline function (most likely `$dateFromString`) then compare as date types.

**Review by Taaku18 — CHANGES_REQUESTED** [review]

File: bot.py (hunk 3 above)
Comment (Cordila) [reply bot.py]: The string compare does work as its comparing both of them as strings. The issue there is a database integrity issue where we are storing date time as a string, this would require a overhaul of the existing databases

**Comment by Cordila** [comment]
All requested changes have been done and tested.

File: core/config_help.json [hunk 4]
```diff
@@ -373,6 +373,17 @@
       "To disable thread cooldown, do `{prefix}config del thread_cooldown`."
     ]
   },
+  "log_expiration": {
+    "default": "Never",
+    "description": "The duration closed threads will be stored within the database before deletion. Logs that have been closed for longer than this duration will be deleted automatically.",
+    "examples": [
+      "`{prefix}config set thread_cooldown P12DT3H` (stands for 12 days and 3 hours in [ISO-8601 Duration Format](https://en.wikipedia.org/wiki/ISO_8601#Durations))",
+      "`{prefix}config set thread_cooldown 3 days and 5 hours` (accepted readable time)"
```
Comment (Jerrie-Aries) [inline core/config_help.json:381]:
```suggestion
      "`{prefix}config set log_expiration P12DT3H` (stands for 12 days and 3 hours in [ISO-8601 Duration Format](https://en.wikipedia.org/wiki/ISO_8601#Durations))",
      "`{prefix}config set log_expiration 3 days and 5 hours` (accepted readable time)"
```

File: bot.py [hunk 5]
```diff
@@ -1697,6 +1695,18 @@ async def before_autoupdate(self):
             self.autoupdate.cancel()
             return
 
+    @tasks.loop(hours=1, reconnect=False)
+    async def log_expiry(self):
+        log_expire_after = self.config.get("log_expiration")
+        if log_expire_after == isodate.Duration():
+            return self.log_expiry.stop()
```
Comment (Jerrie-Aries) [inline bot.py:1702]: What about just doing `return` here without stopping the loop?

That way if we set the `log_expiration` later, the loop will just continue to execute the next block without us needing to restart the bot.

**Review by Jerrie-Aries — COMMENTED** [review]
There were copy and paste errors in `config_help.json`, I believe.

Also a suggestion about not stopping the loop if the config for it is not set.

File: bot.py (hunk 5 above)
Comment (Cordila) [reply bot.py:1702]: The reason why I stopped the loop instead of just returning is that the restart serves as a failsafe mechanism if someone accidentally sets a log expiration, and it just so happens to loop right before they can change it. An edge case scenario really.

**Comment by Jerrie-Aries** [comment]
I did a couple of quick tests. Seems working fine for me.

**Review by Taaku18 — APPROVED** [review]
Thank you for making the requested changes! Looks good.
